Add fire blending to GOCART

bbakernoaa opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
<!-- This description will become the commit message for the PR-->
<!-- Please use this format for your description:

  Describe your changes. Focus on the *what* and *why*. The *how* will be evident from the changes. In particular, be sure to note any interface changes, such as command line syntax, that will need to be communicated to users.

  At the end of your description, please be sure to add the issue this PR solves using the word "Resolves". If there are any issues that are related but not yet resolved (including in other repos), you may use "Refs".

  Resolves #1234
  Refs #4321
  Refs NOAA-EMC/repo#5678
-->

Depends on #3394

# Type of change
<!-- Delete all except one -->
- Bug fix (fixes something broken)
- New feature (adds functionality)
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES/NO
- Does this change require a documentation update? YES/NO
- Does this change require an update to any of the following submodules? YES/NO (If YES, please add a link to any PRs that are pending.)
  - [ ] EMC verif-global
  - [ ] GDAS
  - [ ] GFS-utils
  - [ ] GSI
  - [ ] GSI-monitor
  - [ ] GSI-utils
  - [ ] UFS-utils
  - [ ] UFS-weather-model
  - [ ] wxflow


# How has this been tested?
<!-- Please list any test you conducted, including the machine.

Example:
- Clone and build on WCOSS
- Cycled test on Orion
- Forecast-only on Hera
-->

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

bbakernoaa committed:
add bones to the fire blending in the prep_emissions job

bbakernoaa committed:
pycodestyle

bbakernoaa committed:
pycodestyle fixes

bbakernoaa committed:
fix prep_emission jjob

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

aerorahul committed:
Merge branch 'develop' into feature/fire_blending

aerorahul commented on ush/compare_f90nml.py:
```diff
@@ -78,7 +78,7 @@ def _print_diffs(diff_dict: Dict) -> None:
                 items = diff_dict[path][kk]
                 print(
-                    f"{kk:>{max_len+2}} : {' | '.join(map(str, diff_dict[path][kk]))}")
+                    f"{kk:>{max_len + 2}} : {' | '.join(map(str, diff_dict[path][kk]))}")
```
@WalterKolczynski-NOAA 
Any idea why the pynorms test did not catch this (and ones like this) before?

WalterKolczynski-NOAA replied:
Nope. Maybe it ignores it in f-strings (or otherwise has trouble parsing them)?

aerorahul replied:
not sure why it started picking it up now.

AntonMFernando-NOAA committed:
Cleanup job for GEFS (#2919)

As forecast ensemble jobs are added to the global workflow, this PR
ensures the output is being cleaned up properly once it is no longer
needed.
  Resolves #833

BoCui-NOAA committed:
Update config.resources for bufr sounding job postsnd (#2917)

This PR updates the parm/config/gfs/config.resources and env/WCOSS2.env
files for the BUFR sounding job "postsnd." It includes adjustments to
resource settings such as tasks per node and memory allocations for
various GFS resolutions, including C768, C1152, and others.

Here are the proposed changes:

C768: 7 nodes, 21 tasks per node
C1152: 16 nodes, 9 tasks per node

mingshichen-noaa committed:
Update global atmos upp job to use COMIN/COMOUT (#2867)

NCO has requested that each COM variable specify whether it is an input
or an output. This completes that process for the global-workflow
Unified Post Processor (UPP) task.
Refs: https://github.com/NOAA-EMC/global-workflow/issues/2451

KateFriedman-NOAA committed:
Update to obsproc/v1.2.0 and prepobs/v1.1.0 (#2903)

This PR updates the `develop` branch to use the newer operational
`obsproc/v1.2.0` and `prepobs/v1.1.0`. The obsproc/prepobs installs in
glopara space on supported platforms use tags cut from the `dev/gfsv17`
branches in the respective repos. The installation of `prepobs/v1.1.0`
on WCOSS2 is called "gfsv17_v1.1.0" to help avoid GFSv16 users using it
instead of the operational module.

Also, the `HOMEobsproc` path is updated to set an empty default for
`obsproc_run_ver`. This both removes the need to set a default (and
constantly update it, which is duplication) and avoid the unset variable
error when the fcst jobs use their own load module script that does not
know `obsproc_run_ver`:
```
export HOMEobsproc="${BASE_GIT:-}/obsproc/v${obsproc_run_ver:-}"
```

This PR also reverts the prepobs and fit2obs installs on MSU back to the
glopara space from the temporary `/work/noaa/global/kfriedma/glopara`
space installs.

Lastly, this PR also includes updates to complete issue #2844 (merge
`build.spack.ver` and `run.spack.ver`).

Resolves #2291
Resolves #2840
Resolves #2844

bbakernoaa committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/fire_blending

aerorahul committed:
Merge branch 'develop' into feature/fire_blending

aerorahul commented on parm/config/gefs/config.prep_emissions:
```diff
@@ -2,10 +2,38 @@
 
 ########## config.prep_emissions ##########
 # aerosol emissions preprocessing specific
+# TODO: this is duplicated in the config.aero file.  Should use a common function
+case ${machine} in
+  "HERA")
+    AERO_INPUTS_DIR="/scratch1/NCEPDEV/global/glopara/data/gocart_emissions"
+    ;;
+  "ORION" | "HERCULES")
+    AERO_INPUTS_DIR="/work2/noaa/global/wkolczyn/noscrub/global-workflow/gocart_emissions"
+    ;;
+  "S4")
+    AERO_INPUTS_DIR="/data/prod/glopara/gocart_emissions"
+    ;;
+  "WCOSS2")
+    AERO_INPUTS_DIR="/lfs/h2/emc/global/noscrub/emc.global/data/gocart_emissions"
+    ;;
+  "GAEA")
+    AERO_INPUTS_DIR="/gpfs/f5/epic/proj-shared/global/glopara/data/gocart_emissions"
+    ;;
+  "JET")
+    AERO_INPUTS_DIR="/lfs4/HFIP/hfv3gfs/glopara/data/gocart_emissions"
+    ;;
+  *)
+    echo "FATAL ERROR: Machine ${machine} unsupported for aerosols"
+    exit 2
+    ;;
+esac
+export AERO_INPUTS_DIR
```
I think we can/shouldplace `AERO_INPUTS_DIR` in `workflow/hosts/<machine>.yaml` and then here, replace this section with:
```bash
export AERO_INPUTS_DIR="@AERO_INPUTS_DIR@"
```
Repeat the same in `config.aero`.
What do you think @WalterKolczynski-NOAA ?

ps. this can be done in a follow-up PR.

bbakernoaa replied:
Agree. Will work on a separate PR after this one

WalterKolczynski-NOAA replied:
Agree with Rahul. Moving the paths to the hosts files is the way to go.

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,24 @@
+    qfed_version: '006'                            # qfed version
+    species: [ 'so2','oc','bc' ]                   # species to be used
+    historical: False                              # set to true to just use true data for the given day
+    data:
```
```suggestion
    data_in:
```

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,24 @@
+        {% endfor %}
+    data_out:
+      copy:
+        - ["{{ DATA }}/gefs_blended_emissions.{{ current_cycle | strftime('%Y%m%d') }}.nc", "{{ COMOUT_CHEM_HISTORY }}/{{ RUN }}.{{ current_cycle | strftime('%Y%m%d') }}.gefs_blended_emissions.nc"]
```
```suggestion
        - ["{{ DATA }}/{{ RUN }}_blended_emissions.{{ current_cycle | to_YMD }}.nc", "{{ COMOUT_CHEM_HISTORY }}/{{ RUN }}.{{ current_cycle | to_YMD }}.{{ RUN }}_blended_emissions.nc"]
```

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,24 @@
+        - "{{ DATA }}"
+        - "{{ COMOUT_CHEM_HISTORY}}"
+      copy:
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_oc.006.{{ current_cycle | strftime('%Y%m%d') }}.nc4", "{{ DATA }}/"]
```
```suggestion
        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_oc.006.{{ current_cycle | to_YMD }}.nc4", "{{ DATA }}/"]
```

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,24 @@
+        - "{{ COMOUT_CHEM_HISTORY}}"
+      copy:
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_oc.006.{{ current_cycle | strftime('%Y%m%d') }}.nc4", "{{ DATA }}/"]
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_so2.006.{{ current_cycle | strftime('%Y%m%d') }}.nc4", "{{ DATA }}/"]
```
```suggestion
        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_so2.006.{{ current_cycle | to_YMD }}.nc4", "{{ DATA }}/"]
```

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,24 @@
+      copy:
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_oc.006.{{ current_cycle | strftime('%Y%m%d') }}.nc4", "{{ DATA }}/"]
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_so2.006.{{ current_cycle | strftime('%Y%m%d') }}.nc4", "{{ DATA }}/"]
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_bc.006.{{ current_cycle | strftime('%Y%m%d') }}.nc4", "{{ DATA }}/"]
```
```suggestion
        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_bc.006.{{ current_cycle | to_YMD }}.nc4", "{{ DATA }}/"]
```

aerorahul commented on parm/ufs/gocart/ExtData.hfed:
```diff
@@ -0,0 +1,8 @@
+#====== BIOMASS BURNING EMISSIONS =======================================
+
+# HFED
+#--------------------------------------------------------------------------------------------------------------------------------
+SU_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none 0.7778 biomass ExtData/nexus/HFED/Y1994/M%m2/hfed.emis_so2.x576_y361.%y4%m2.nc4
+OC_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none 0.7778 biomass ExtData/nexus/HFED/Y1994/M%m2/hfed.emis_oc.x576_y361.%y4%m2.nc4
+BC_BIOMASS NA  N Y %y4-%m2-%d2t12:00:00 none 0.7778 biomass ExtData/nexus/HFED/Y1994/M%m2/hfed.emis_bc.x576_y361.%y4%m2.nc4
+# EMI_NH3_BB NA  N Y %y4-%m2-%d2t12:00:00 none 0.7778 biomass ExtData/nexus/HFED/Y1994/M%m2/hfed.emis_nh3.x576_y361.%y4%m2.nc4
```
This file should be linked from UFS and not added in the global-workflow.

bbakernoaa replied:
I just mimicked what was done for the other fire emissions.  The ExtData folder is linked by the workflow to the current run directory.  So this just points the the file within that directory structure.

aerorahul replied:
ok. we can address this at a later time.

aerorahul commented on scripts/exglobal_prep_emissions.py:
```diff
@@ -19,7 +29,17 @@
-    emissions.finalize()
+
+    # get local keys for configuration
+    keys = ['DATA', 'forecast_dates', 'cdate', 'aero_emission_yaml']
```
We have removed instances of `cdate`.  Please use `PDY` and `cyc` in its place.  I will try to find these instances as I go through the review and make suggestions.

aerorahul commented on parm/config/gefs/config.prep_emissions:
```diff
@@ -2,10 +2,38 @@
 echo "BEGIN: config.prep_emissions"
 
 # Get task specific resources
+export STARTHOUR=${STARTHOUR:-00}
```
Why not put `STARTHOUR` in the yaml?  Does it change or have a upstream configuration that provides this value?

bbakernoaa replied:
we don't really need the start hour if its included in the `self.task_config.current_cycle`. I was trying to mimic what upp.py did but its a little different application.  Maybe we just pass 0H here for the `to_timedelta`?

aerorahul replied:
I don't see where `STARTHOUR` is used or referenced in this PR.  Do we need it?

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,27 +44,50 @@ def __init__(self, config: Dict[str, Any]) -> None:
 
         localdict = AttrDict(
-            {'variable_used_repeatedly': local_variable}
+            {"cdate": current_datetime, "nforecast_days": nforecast_days, "forecast_dates": forecast_dates}
```
Seems like `cdate` means `start_date` for blending the emissions. 
`cdate` in global-workflow has a meaning of `PDYcyc`.  
I suggest, we call this variable, `blend_start_date` to distinguish it from the `current_cycle = CDATE = PDYcyc`.

aerorahul replied:
Don't see `cdate` used anywhere.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,27 +44,50 @@ def __init__(self, config: Dict[str, Any]) -> None:
+    def initialize(aero_emission_yaml: Dict) -> None:
+        """Initialize the work directory"""
+        logger.info("Copy Static Data to run directory")
+        # FileHandler(aero_emission_yaml.aero_emissions.fix_data).sync()
```
Is this still a `TODO`?

bbakernoaa replied:
The data is copied in the configure step here.  Removing comment

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,27 +44,50 @@ def __init__(self, config: Dict[str, Any]) -> None:
-        """Initialize the work directory
-        """
+    def initialize(aero_emission_yaml: Dict) -> None:
+        """Initialize the work directory"""
```
Should include `numpy` doc block.  See `upp.py` for example.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,27 +44,50 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        logger.info(
+            f"Copy '{aero_emission_yaml.aero_emissions.config.data}' data to run directory"
+        )
+        FileHandler(aero_emission_yaml.aero_emissions.config.data).sync()
```
```suggestion
        FileHandler(aero_emission_yaml.aero_emissions.config.data_in).sync()
```
Why not do this in `initialize`.  It is copying fixed data into the `DATA` directory.

bbakernoaa replied:
we could.  Moving

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        emistype = Config_dict.emistype
+        ratio = Config_dict.ratio
+        climfiles = sort(glob("{}{}".format(Config_dict.climfile_str, "*.nc")))
+        print(Config_dict.data_out['copy'][0][0])
```
Use `logger.info`

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,24 @@
+    debug: False
+    ratio: 0.95                                    # weighting ratio
+    emistype: 'QFED'                               # EMission Type, Valid answers: 'QFED, GBBEPx, HFED'
+    climfile_str: 'GBBEPx-all01GRID_v4r0_climMean' # climate file base string used for glob later
```
Does this `climfile_str` change when `emistype` changes?

bbakernoaa replied:
No it wouldn't.  But we do have a few different options for which climatology we have.  Mean or 95th percentile etc.  So the filename could change to accommodate that

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        climfiles = sort(glob("{}{}".format(Config_dict.climfile_str, "*.nc")))
+        print(Config_dict.data_out['copy'][0][0])
+
+        if emistype.lower() == "qfed":
+            basefile = glob("qfed2.emis_*.nc4")
+        elif emistype.lower() == "gbbepx":
+            basefile = glob("GBBEPx_all01GRID.emissions_v*.nc")
+        elif emistype.lower() == "hfed":
+            basefile = glob("hfed.emis_*.x576_y361.*nc4")
```
I think we can provide the regex string to `glob` in the yaml.
This section will then simply be:
```python
basefile = glob(emis_glob_str)
```

bbakernoaa replied:
I guess I'm a little confused by this because the glob would need to occur after the initialization method due to the files getting synced in that step. So I don't know if we could glob it within the yaml file itself

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+            ratio=ratio,
+            scale_climo=True,
+            obsfile=basefile)
+        print(dset)
```
This should be printed when debugging, and please use `logger.info` or `logger.debug`.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+            Dataset containing the fire emissions data
+        """
+
+        vrs = ["BC", "CH4", "CO", "CO2", "NH3", "NOx", "OC", "PM2.5", "SO2"]
+        qfed_vars = ["bc", "ch4", "co", "co2", "nh3", "no", "oc", "pm25", "so2"]
```
Should be part of the yaml.

bbakernoaa replied:
finshed the input from the yaml file

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+            index_good = [[i, v] for i, v in enumerate(qfed_vars) if v in f]
+            good = index_good[0][0]
+            found_species.append(index_good[0][1])
+            da = xr.open_dataset(f, decode_cf=False).biomass
```
Shouldn't there be a `da.close`?
If possible use:
```python
with xr.open_dataset(f, decode_cf=False).biomass as da:
   da.name = ...
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+
+    @staticmethod
+    @logit(logger)
+    def open_climatology(fname) -> xr.Dataset:
```
need doc-block

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+    @logit(logger)
+    def open_qfed(fname) -> xr.Dataset:
+        """
+        Open QFED2 fire emissions data
```
Open and do what with the data?
read variables, return the dataset of variables are found.  
What happens if none of the species are found?  Will it still return a dataset?  Should there be a warning/fatal error to trap such a condition?  
This should be done here, rather than some other method downstream expecting a dataset and finding empty dset.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        if len(fname) > 1:
+            files = sort(fname)
+        else:
+            files = sort(glob(fname))
```
A `logger.info` message would be nice to report what files it found and it is about to loop through them.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        # print(files)
+        xr.open_dataset(files[0])
+        for i, f in enumerate(files):
+            # print("  opening:", f)
```
A `logger.info` would be nice to have here to indicate while file is being processed.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+            encoding["lon_b"] = dict(zlib=True, complevel=4)
+        if "time" in dset:
+            encoding["time"] = dict(dtype="i4")
+        dset.load().to_netcdf(outfile, encoding=encoding)
```
Use try, except to catch exceptions if something goes wrong
```suggestion
        try:
            dset.load().to_netcdf(outfile, encoding=encoding)
        except Exception as ee:
            logger.exception(f"Encountered an exception in writing the dataset, {ee}")
            raise Exception("FATAL ERROR: Unable to write dataset, ABORT!")
```
xarray.dataset provides numerous Exception handling.  Please look at the documentation and at least try to trap a couple of most commonly occurring exceptions, such as `OSError` - system had a glitch like disk is full, `ValueError` - invalid dataset, etc.
This will help in debugging and tracing issues at the source.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        """
+        Create scaled climatology data based on emission data.
+
+        Parameters:
```
Use numpy documentation format

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+    @staticmethod
+    @logit(logger)
+    def create_climatology(
+        emissions, climatology, lat_coarse=50, lon_coarse=50
```
Use type hints please.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+    @staticmethod
+    @logit(logger)
+    def make_fire_emission(
+        d=None,
```
use type hints

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        Returns:
+        - list: A list of xarray.Dataset objects representing fire emissions data for each forecast day.
+        """
+        # import pandas as pd
+        import numpy as np
```
Imports should be at the top of the module

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        # dd = pd.Timestamp(d)
+
+        # open fire emission
+        if len(obsfile) > 1:
+            if "QFED".lower() in obsfile[0].lower():
+                g = AerosolEmissions.open_qfed(obsfile)
+            else:
+                g = xr.open_mfdataset(obsfile, decode_cf=False)
+        else:
+            if "QFED".lower() in obsfile.lower():
+                g = AerosolEmissions.open_qfed(obsfile)
+            else:
+                g = xr.open_mfdataset(obsfile, decode_cf=False)
```
Why not make obsfile a list, and avoid this duplication.
```suggestion
        if isinstance(obsfile, (str, bytes)):
            obsfile = [obsfile]
        if "QFED".lower() in obsfile[0].lower():
            g = AerosolEmissions.open_qfed(obsfile)
        else:
            g = xr.open_mfdataset(obsfile, decode_cf=False)
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        climo = climo.sel(lat=g["lat"], lon=g["lon"], method="nearest")
+
+        # make weighted climo
+        gc = g.coarsen(lat=150, lon=150, boundary="trim").sum()
```
Should these coarse grid values be configurable in the yaml?

bbakernoaa replied:
Yes they can be.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        dsets = []
+        climo_scaled = {}
+        for tslice in np.arange(len(climos)):
+            print(tslice)
```
Use logger.info, pleae

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
         """Perform closing actions of the task.
         Copy data back from the DATA/ directory to COM/
         """
+        print(Config_dict.data_out)
```
`logger.info`

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -75,10 +105,261 @@ def execute(workdir: Union[str, os.PathLike], aprun_cmd: str) -> None:
+        """
+        Generate fire emissions data for a given date and forecast period.
+
+        Parameters:
```
Please use `numpy` documentation formatting.

aerorahul reviewed (changes requested):
@bbakernoaa 
Please let me know if any of my comments need clarifications

bbakernoaa committed:
Update parm/prep/aero_emissions.yaml

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update parm/prep/aero_emissions.yaml

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update parm/prep/aero_emissions.yaml

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update parm/prep/aero_emissions.yaml

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
add coarsen_scale to yaml and use in aero_emission.py

bbakernoaa committed:
remove debugging print statements

bbakernoaa committed:
Update parm/prep/aero_emissions.yaml

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
add init doc block

bbakernoaa committed:
remove comment

bbakernoaa committed:
move syncing to initialize

bbakernoaa committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/fire_blending' into feature/fire_blending

bbakernoaa committed:
add loger info for this and error trapping

bbakernoaa committed:
add more changes

bbakernoaa committed:
more changes, docstrings, error checking etc

bbakernoaa committed:
pycodestyle changes

bbakernoaa committed:
add blended emission option

bbakernoaa committed:
copy blending from COMOUT if AERO_EMIS_FIRE == blending

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -675,13 +675,19 @@ GOCART_rc() {
 
+  # Link blending emissions if AERO_EMIS_FIRE == blending
+  if [[ "${AERO_EMIS_FIRE}" == 'blending' ]]; then 
+    ${NCP} "${COMOUT_CHEM_HISTORY}/gefs.${vdate:0:8}.gefs.blended_emissions.nc" "${DATA}"
```
```suggestion
    ${NCP} "${COMOUT_CHEM_HISTORY}/${RUN}.${vdate:0:8}.${RUN}.blended_emissions.nc" "${DATA}"
```

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -675,13 +675,19 @@ GOCART_rc() {
   fi
 
+  # Link blending emissions if AERO_EMIS_FIRE == blending
+  if [[ "${AERO_EMIS_FIRE}" == 'blending' ]]; then 
```
I think this is only relevant for `RUN=gefs`, so we should add that conditionality. Atleast till such time that `RUN=gfs` also runs w/ aerosols
```suggestion
  if [[ "${AERO_EMIS_FIRE}" == "blending" && "${RUN}" == "gefs" ]]; then 
```

CoryMartin-NOAA replied:
GDAS_fcst will run with aerosols

aerorahul replied:
Thanks @CoryMartin-NOAA 
Will `GDAS_fcst` use blended emissions?  Presently, this task is currently only activated for GEFS (and SFS) -- at least that is what was communicated.

CoryMartin-NOAA replied:
oh, I don't know, @bbakernoaa any thoughts?

bbakernoaa replied:
If the GDAS forecast is less than a day or two we definitely don't need this at all.  The current setup is to only add the blending after 5 days of assumed persistence.

CoryMartin-NOAA replied:
9 hours, max, so we don't need it for GDAS then. Thanks

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
-        local_variable = "something"
+        current_datetime = add_to_datetime(
+            self.task_config.PDY,
+            to_timedelta(f"0H"),
```
Guessing this is where `STARTHOUR` is relevent?

bbakernoaa replied:
Yes but it should always be 0H here as the emissions are daily.  I'll remove `STARTHOUR` from the config

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        elif emistype.lower() == "gbbepx":
+            basefile = glob("GBBEPx_all01GRID.emissions_v*.nc")
+        elif emistype.lower() == "hfed":
+            basefile = glob("hfed.emis_*.x576_y361.*nc4")
```
```suggestion
        emission_map = {'qfed': 'qfed2.emis_*.nc4',
            'gbbepx': 'GBBEPx_all01GRID.emissions_v*.nc',
            'hfed': 'hfed.emis_*.x576_y361.*nc4'}
        try:
            basefile = glob(emission_map[emistype.lower()])
        except KeyError as err:
            raise KeyError(f"FATAL ERROR: {emistype.lower()} is not a supported emission type, ABORT!")
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        if "latitude" in dset:
+            encoding["latitude"] = dict(zlib=True, complevel=4)
+            encoding["longitude"] = dict(zlib=True, complevel=4)
+        if "lat_b" in dset:
```
Is there a possibility that `lat_b` and `latitude` are both in this file?  If so, what happens then.

bbakernoaa replied:
its actually not.  I copied this little write utility from something else I had.  But the answer is nothing. This just compresses the data using the zlib.  It doesn't modify any variable names etc.

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        """
+        Generate fire emissions data for a given date and forecast period.
+
+        Parameters:
```
```suggestion
        Parameters:
        -----------
```

aerorahul reviewed (commented):
some comments after your recent push.
Nothing new, just following up on previous review.

bbakernoaa committed:
Update ush/forecast_postdet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update ush/forecast_postdet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
changing resources

bbakernoaa committed:
remove starthour

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

bbakernoaa committed:
fix pycodestyle

bbakernoaa committed:
complete passing vars from yaml

bbakernoaa committed:
pycodestyle again

bbakernoaa committed:
fix passing matching vars arrays from yaml file

bbakernoaa committed:
more changes to fix issues

aerorahul commented on parm/prep/aero_emissions.yaml:
```diff
@@ -0,0 +1,27 @@
+        - "{{ DATA }}"
+        - "{{ COMOUT_CHEM_HISTORY}}"
+      copy:
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_oc.006.{{ current_cycle | to_YMD }}.nc4", "{{ DATA }}/"]
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_so2.006.{{ current_cycle | to_YMD }}.nc4", "{{ DATA }}/"]
+        - ["{{ AERO_INPUTS_DIR }}/nexus/QFED/{{ current_cycle | strftime('%Y/%m') }}/qfed2.emis_bc.006.{{ current_cycle | to_YMD }}.nc4", "{{ DATA }}/"]
+        {% for fdate in forecast_dates %}
+        - ["{{ AERO_INPUTS_DIR }}/nexus/GBBEPx/v4/climMean/GBBEPx-all01GRID_v4r0_climMean_{{ fdate | strftime('%m%d') }}.nc", "{{ DATA }}/"] # copy climo files
```
What happens when `emistype` is not `QFED`? Are these still the relevant files to be copied?

bbakernoaa replied:
if the emission type changes then we need to modify these lines.  I couldn't figure out how to put a condition in for the different emission types.  If you have a suggestion on how we can do this in jinja I'm happy to add it.  it will make everyone's lives easier.

aerorahul replied:
One idea, that is just simple, is:
```yaml
data_in:
    qfed:
         copy:
            -
            -
    hfed:
        copy:
            -
            -
```
Handle the `mkdir` part separately from `copy` since that is common.
Just a thought
You get everything right in front in the yaml

bbakernoaa replied:
I'll give it a shot

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -4,23 +4,31 @@
 from typing import Dict, Any, Union
 from pprint import pformat
+from glob import glob
+from numpy import sort
+import numpy as np
```
```suggestion
import numpy as np
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        """
+        emistype = Config_dict.emistype
+        ratio = Config_dict.ratio
+        climfiles = sort(glob("{}{}".format(Config_dict.climfile_str, "*.nc")))
```
```suggestion
        climfiles = np.sort(glob("{}{}".format(Config_dict.climfile_str, "*.nc")))
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        qfed_vars = input_vars  # ["bc", "ch4", "co", "co2", "nh3", "no", "oc", "pm25", "so2"]
+
+        if len(fname) > 1:
+            files = sort(fname)
+        else:
+            files = sort(glob(fname))
```
```suggestion
            files = np.sort(fname)
        else:
            files = np.sort(glob(fname))
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,383 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        das = []
+
+        if len(fname) > 1:
+            files = sort(fname)
+        else:
+            files = sort(glob(fname))
```
```suggestion
            files = np.sort(fname)
        else:
            files = np.sort(glob(fname))
```

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
add changes to copy the data for each fire case and add a method that if hfed is used just to transform the data... ie don't really process

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -85,10 +85,32 @@ def initialize(aero_emission_yaml: Dict) -> None:
+                f"Copy Climatology '{aero_emission_yaml.data_in.climo}' data to run directory"
+            )
+            FileHandler(aero_emission_yaml.data_in.climo).sync()
+        if aero_emission_yaml.emistype.lower() == 'qfed':
+            # Copy QFED and climatology
+            logger.info(
+                f"Copy QFED '{aero_emission_yaml.data_in.qfed}' data to run directory"
+            )
+            FileHandler(aero_emission_yaml.data_in.qfed).sync()
+        elif aero_emission_yaml.emistype.lower() == 'gbbepx':
+            # Copy GBBEPx and climatology
+            logger.info(
+                f"Copy GBBEPx '{aero_emission_yaml.data_in.gbbepx}' data to run directory"
+            )
+            FileHandler(aero_emission_yaml.data_in.gbbepx).sync()
+        elif aero_emission_yaml.emistype.lower() == 'hfed':
+            # Copy HFED and climatology
+            logger.info(
+                f"Copy HFED '{aero_emission_yaml.data_in.hfed}' data to run directory"
+            )
+            FileHandler(aero_emission_yaml.data_in.hfed).sync()
```
```suggestion
        logger.info(f"Copy {aero_emission_yaml.emistype} data to run directory")
        FileHandler(aero_emission_yaml.data_in[aero_emission_yaml.emistype.lower()]).sync()
```

aerorahul commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -135,17 +157,70 @@ def run(cls, workdir: Union[str, os.PathLike], current_date: str = None, forecas
+
+        try:
+            if np.sort(files).size == 0:
+                raise Exception
```
```suggestion
                raise Exception("FATAL ERROR: describe the error ...")
```

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa committed:
last suggetions?

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

bbakernoaa committed:
add with command when opening files

bbakernoaa committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/fire_blending' into feature/fire_blending

DavidHuber-NOAA commented on scripts/exglobal_prep_emissions.py:
```diff
@@ -19,7 +29,18 @@
+        edict[key] = emissions.task_config[key]
+    edict['CONFIG'] = edict.aero_emission_yaml.aero_emissions['config']
+    edict.aero_emission_yaml['emistype'] = edict['CONFIG'].emistype
+    print(edict)
```
```suggestion
```

DavidHuber-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,446 @@ def __init__(self, config: Dict[str, Any]) -> None:
         """
+        logger.info("Copy Static Data to run directory")
+
+        print(aero_emission_yaml)
```
```suggestion
```

Or make it into a logger entry.

bbakernoaa replied:
sorry was debugging things and forgot to remove this

bbakernoaa committed:
Update scripts/exglobal_prep_emissions.py

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

DavidHuber-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,446 @@ def __init__(self, config: Dict[str, Any]) -> None:
         super().__init__(config)
 
-        local_variable = "something"
+        current_datetime = add_to_datetime(
+            self.task_config.PDY,
+            to_timedelta(f"0H"),
+        )
```
Just to be sure, is the intent here to set `current_datetime` to 00Z on `PDY`?  If so, I think you can simplify this to just

```suggestion
        current_datetime = self.task_config.PDY
```

DavidHuber-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,446 @@ def __init__(self, config: Dict[str, Any]) -> None:
         """
+        logger.info("Copy Static Data to run directory")
+
+        print(aero_emission_yaml)
```
Suggest either deleting or changing to a debug logger.

bbakernoaa replied:
We should just remove it.

DavidHuber-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,446 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        invars = input_vars
+
+        try:
+            if np.sort(files).size == 0:
```
`len` is a valid function to run on numpy arrays:

```suggestion
            if len(files) == 0:
```

DavidHuber-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,446 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            clim_slice = clim.data[index, :, :]
+
+            # Calculate the weighted alpha ratio parameter
+            # alpha = 1.0 - 1.0 / (index + 1)
```
Is this an option for a later time, an experimental option, or an interpretation of line 378? If the former, should it have a `TODO`?

bbakernoaa replied:
Can remove it from here. This script was originally developed outside of the workflow and has some legacy items there

aerorahul commented:
@bbakernoaa 
Others, please don't hold this PR for my review.  I think I have reviewed this PR and will let others review and provide comments.

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

bbakernoaa committed:
Update aero_emissions.py

resolving comments

WalterKolczynski-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -675,13 +675,19 @@ GOCART_rc() {
   fi
 
+  # Link blending emissions if AERO_EMIS_FIRE == blending
+  if [[ "${AERO_EMIS_FIRE}" == "blending" && "${RUN}" == "gefs" ]]; then 
```
Is there a reason for the check against `$RUN`? Presumably, if `$AERO_EMIS_FIRE` is blending, we want to use blending regardless of the `$RUN`.

bbakernoaa replied:
I'm not sure... Rahul asked to add this but

aerorahul replied:
Since this was only GEFS related, I suggested the `RUN` condition check.  Pretty sure, SFS will need that as well in the future.
I can go either way.

WalterKolczynski-NOAA replied:
I'd rather rely on the app config to know whether it wants blending.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        current_datetime = self.task_config.PDY
+        nforecast_hours = self.task_config["FHMAX_GFS"]
+        nforecast_days = nforecast_hours // 24 + 1
+        forecast_dates = [current_datetime + timedelta(days=i) for i in range(nforecast_days)]
```
This might be clearer. It should produce the same list as the existing code.
```
from rrule import rrule, DAILY
```
```suggestion
        nforecast_hours = self.task_config["FHMAX_GFS"]
        blend_start_date = self.task_config.PDY
        blend_end_date = blend_start_date + to_timedelta(f'{nforecast_hours+24}H')
        forecast_dates = list(rrule(freq=DAILY, dtstart=blend_start_date, until=blend_end_date))
```

WalterKolczynski-NOAA commented on scripts/exglobal_prep_emissions.py:
```diff
@@ -19,7 +29,17 @@
-    emissions.execute(emissions.task_config.DATA, emissions.task_config.APRUN)
-    emissions.finalize()
+
+    # get local keys for configuration
+    keys = ['DATA', 'forecast_dates', 'PDY', 'cyc', 'aero_emission_yaml']
+    edict = AttrDict()
+    for key in keys:
+        edict[key] = emissions.task_config[key]
+    edict['CONFIG'] = edict.aero_emission_yaml.aero_emissions['config']
+    edict.aero_emission_yaml['emistype'] = edict['CONFIG'].emistype
```
This information should all stay inside the class. `AerosolEmissions.initialize()`, `AerosolEmissions.run()`, etc. should all be regular methods rather than static or class methods, with all this information kept within the object.

WalterKolczynski-NOAA replied:
I see you followed how upp did it. I'm not sure why it takes the data out of the class to put it back in either. Compare against what snow_analysis.py is doing. I'm going to make a note to refactor upp as well.

WalterKolczynski-NOAA replied:
If this was something you worked through with Rahul, let me know and we'll proceed with it as-is.

bbakernoaa replied:
I was following Rahul's suggestion and following the upp.py example.  I can try to make this change as I agree it is a little wonky like this.

WalterKolczynski-NOAA commented on scripts/exglobal_prep_emissions.py:
```diff
@@ -19,7 +29,17 @@
+    emissions.initialize(edict.CONFIG)
+    emissions.configure(edict.aero_emission_yaml)
+    emissions.run(workdir=edict.DATA, current_date=edict.PDY, forecast_dates=edict.forecast_dates, Config_dict=edict.CONFIG)
+    emissions.finalize(edict['CONFIG'])
```
Per above:
```suggestion
    emissions.initialize()
    emissions.configure()
    emissions.run()
    emissions.finalize()
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
-    def configure() -> None:
+    def configure(aero_emission_yaml: Dict) -> None:
         """Configure the artifacts in the work directory.
         Copy run specific data to run directory
```
This method still isn't doing anything.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+                raise Exception("FATAL ERROR: No HFED files were found")
+        except Exception as ee:
+            logger.exception("FATAL ERROR: HFED files not found.")
+            raise Exception(f"FATAL ERROR: Unable to find files, ABORT")
```
I don't think this outer try is needed. No files are being read here, just the length of the list is checked. Also changed the error message since this method isn't finding files, just receiving a list.
```suggestion
        if len(files) == 0:
            raise Exception("FATAL ERROR: Received empty list of HFED files")
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        None
+        """
+        vrs = out_vars
+        invars = input_vars
```
Why change the variable names? This just makes it harder to follow.

zmoon replied:
Addressed in  d8b6d291a584

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+                logger.info("Opening HFED file: {filename}".format(filename=f))
+                with xr.open_dataset(f, decode_cf=False).biomass as da:
+                    da.name = vrs[good]
+                    dset_dict[vrs[good]] = da
```
This bit with the enumerate is non-intuitive to parse. It looks like the purpose is to map input variables to output variables? If so, I would suggest replacing input_vars and out_vars with a dictionary with the mapping and removing this bit with enumerate.

```suggestion
        for filename in files:
            # Even better if the filenames are regular and we can replace this with a regex
            in_var = [var for var in in_vars if var in filename][0]
            out_var = variable_mapping[in_var]
            found_species.append(in_var)
            
            try:
                logger.info(f"Opening HFED file: {filename}")
                with xr.open_dataset(filename, decode_cf=False).biomass as da:
                    da.name = out_var
                    dset_dict[out_var] = da
```

zmoon replied:
Attempted to address in 7ea3b689450, let me know what you think.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+                raise Exception("FATAL ERROR: No QFED files were found")
+        except Exception as ee:
+            logger.exception("FATAL ERROR: QFED files not found.")
+            raise Exception(f"FATAL ERROR: Unable to find files, ABORT")
```
```suggestion
        if len(files) == 0:
            raise Exception("FATAL ERROR: Received empty list of QFED files")
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+                    dset_dict[vrs[good]] = da
+            except Exception as ee:
+                logger.exception("FATAL ERROR: unable to read dataset {error}".format(error=ee))
+                raise Exception("FATAL ERROR: Unable to read dataset, ABORT!")
```
Change similarly to above.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        if len(fname) > 1:
+            files = np.sort(fname)
+        else:
+            files = np.sort(glob(fname))
```
We would prefer to avoid globs if possible, since they are a bit unsafe. Is this option needed?

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        if len(fname) > 1:
+            files = np.sort(fname)
+        else:
+            files = np.sort(glob(fname))
```
See above

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        logger.info("Process Climatlogy Files")
+        logger.info("  Opening Climatology File: {filename}".format(filename=fname[0]))
+        xr.open_dataset(files[0])
+        for i, f in enumerate(files):
```
The index isn't being used, I don't think you need enumerate
```suggestion
        for filename in files:
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        logger.info("  Opening Climatology File: {filename}".format(filename=fname[0]))
+        xr.open_dataset(files[0])
+        for i, f in enumerate(files):
+            logger.info("  Opening Climatology File: {filename}".format(filename=f))
```
```suggestion
            logger.info(f"  Opening Climatology File: {filename}")
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        ratio_interp = ratio.sel(lat=clim.lat, lon=clim.lon, method="nearest")
+
+        # Loop through each time slice and scale the climatology
+        for index, time_slice in enumerate(clim.time):
```
time_slice isn't needed, so no need to use enumerate
```suggestion
        for index in range(0, clim.time):
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        ratio_interp = ratio.sel(lat=clim.lat, lon=clim.lon, method="nearest")
+
+        # Loop through each time slice and scale the climatology
+        for index, time_slice in enumerate(clim.time):
+            # Get the current time slice of the climatology
+            clim_slice = clim.data[index, :, :]
+
+            # Scale the current time slice
+            scaled_slice = clim_slice * ratio_interp[index, :, :]
+
+            # Update the climatology with the scaled time slice
+            clim.data[index, :, :] = scaled_slice.squeeze().data
```
Does this even need to be a loop, or can we just do:
```suggestion
        # Scale data
        clim.data = clim.data * ratio_interp
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        if isinstance(obsfile, (str, bytes)):
+            obsfile = [obsfile]
+        if "QFED".lower() in obsfile[0].lower():
+            g = AerosolEmissions.open_qfed(obsfile, out_vars=output_vars, input_vars=input_vars)
```
Please use descriptive variable names instead of single letters.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +44,439 @@ def __init__(self, config: Dict[str, Any]) -> None:
+
+            for v in g.data_vars:
+                if not scale_climo:
+                    if tslice > 5:
```
What's special about 5? Should this be some sort of variable instead of a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants)?

bbakernoaa replied:
Magic numbers are fun.  This is honestly an empirical number that represents the amount of days that we could potentially assume persistance.  We could make it a variable though to give more flexibility

WalterKolczynski-NOAA replied:
Please do.

WalterKolczynski-NOAA reviewed (changes requested):
Let me know if you have questions about the refactoring to keep the data inside the class (or anything else).

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

bbakernoaa committed:
addressing many issues

bbakernoaa committed:
Adding n_persist to the yaml file and populating that through

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -682,13 +682,19 @@ GOCART_rc() {
 
+  # Link blending emissions if AERO_EMIS_FIRE == blending
+  if [[ "${AERO_EMIS_FIRE}" == "blending" && "${RUN}" == "gefs" ]]; then 
+    ${NCP} "${COMOUT_CHEM_HISTORY}/${RUN}.${vdate:0:8}.${RUN}.blended_emissions.nc" "${DATA}"
```
Where is `vdate` coming from in this function scope?

zmoon replied:
I updated it to use `$current_cycle` instead, consistent with the `aero_emissions.yaml`. Is that ok?

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,443 @@ def __init__(self, config: Dict[str, Any]) -> None:
 
-        local_variable = "something"
+        nforecast_hours = self.task_config["FHMAX_GFS"]
+        blend_start_date = self.task_config.PDY
```
This is my fault, I should've been consistent when I made the suggestion:
```suggestion
        nforecast_hours = self.task_config["FHMAX_GFS"]
        blend_start_date = self.task_config["PDY"]
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,443 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            {
+                "cdate": blend_start_date,
+                "nforecast_days": nforecast_hours // 24,
+                # "forecast_dates": forecast_dates,
```
```suggestion
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,443 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        --------
+        None
+        """
+        Config_dict = self.task_config['config']
```
Object instances should be lowercase.
```suggestion
        config_dict = self.task_config['config']
```

WalterKolczynski-NOAA replied:
This line got changed, but the subsequent references to it weren't.

zmoon replied:
Addressed in 5cdd6a60b

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,443 @@ def __init__(self, config: Dict[str, Any]) -> None:
+
+        Parameters:
+        -----------
+        workdir : Union[str, os.PathLike])
+            The working directory path.
+        current_date: (str, optional)
+            The current date for fire emissions generation.
+        forecast_dates: list, optional
+            List of forecast dates.
+        Config_dict:  Dict
+            Configuration dictionary for the task.
```
Not passing in any parameters anymore.
```suggestion
        None
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +45,443 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            for v in ObsEmis.data_vars:
+                if not scale_climo:
+                    if tslice > n_persist:
+                        # kk = ratio * dset[v] + (1 - ratio) * climo[v].data[tslice, :, :]
```
```suggestion
```

WalterKolczynski-NOAA reviewed (changes requested):
Mostly minor stuff in the second pass, but there are a couple suggestions from my previous review that were marked as resolved without being addressed (I marked them back as unresolved).

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

bbakernoaa committed:
Update ush/python/pygfs/task/aero_emissions.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

zmoon committed:
Use current_cycle

zmoon committed:
Seems like this should be underscore

zmoon committed:
`config_dict`

zmoon committed:
Break long line

zmoon committed:
Just use the args

zmoon committed:
Add missing inits

zmoon committed:
Remove some unused imports and sort

zmoon committed:
Address unused vars

zmoon committed:
Seems like these are really assuming lists of str

zmoon committed:
Rename not needed

zmoon committed:
Replace input/output var lists with mapping

zmoon committed:
Some spelling/typing/docstring fixes

zmoon committed:
Merge remote-tracking branch 'origin/develop' into feature/fire_blending

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        if emistype.lower() != 'hfed':
+            logger.info(
+                f"Copy HFED '{data_in.hfed}' data to run directory"
+            )
```
Not sure exactly what is going on here. The log message seems to be the opposite what the if condition is, and the if-block isn't actually doing any action. Is the rest of the method supposed to be in this if-block? If so, what is this first logger message?

zmoon replied:
Addressed in ca3bc1a7f7

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        qfedfiles = [os.path.basename(fname[0]) for fname in config['data_in']['qfed']['copy']]
+        hfedfiles = [os.path.basename(fname[0]) for fname in config['data_in']['hfed']['copy']]
+        gbbepxfiles = [os.path.basename(fname[0]) for fname in config['data_in']['gbbepx']['copy']]
+        climofiles = [os.path.basename(fname[0]) for fname in config['data_in']['climo']['copy']]
```
Instead of using different variables for these, and then having to maintain a list of acceptable emission types later as `emission_map` in `run()`, why not make this a dict?
```suggestion
        emission_files = {}
        for emission_type in ['qfed', 'hfed', 'gbbepx', 'climo']:
            emission_files[emission_type] = [os.path.basename(fname[0]) for fname in config['data_in'][emission_type]['copy']]
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            try:
+                basefile = emission_map[emistype.lower()]
+            except KeyError as err:
+                raise KeyError(f"FATAL ERROR: {emistype.lower()} is not a supported emission type, ABORT!") from err
```
Following the change above, you no longer need the emission_map:
```suggestion
        if emistype.lower() != 'blended':
            try:
                basefile = self.task_config[emistype.lower()}
            except KeyError as err:
                raise KeyError(f"FATAL ERROR: {emistype.lower()} is not a supported emission type, ABORT!") from err
```

zmoon replied:
Addressed in bc9e8f95e67

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        -------
+        None
+        """
+        if out_var_dict is None:
+            raise Exception("FATAL ERROR: No output variable mapping provided")
+
+        if len(files) == 0:
+            raise Exception("FATAL ERROR: Received empty list of HFED files")
```
I think we want a message to the logger for these exceptions before we throw them.

zmoon replied:
With `logger.info()`?

zmoon replied:
Attempted to address in 3227263

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            _, input_var = os.path.basename(f).split(".")[1].split("_")
+            found_species.append(input_var)
+            try:
+                with xr.open_dataset(f, decode_cf=False).biomass as da:
```
Can you add a comment about what's going on here? It seems like all of the files are using `biomass` as the variable name and the real variable name is in the filename, so we are writing a new NetCDF with everything together and the correct variable names?

zmoon replied:
That's my understanding, yes.

zmoon replied:
Added some more info to the docstring in ca3bc1a7f7

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            raise Exception("FATAL ERROR: No output variable mapping provided")
+
+        if len(files) == 0:
+            raise Exception("FATAL ERROR: Received empty list of QFED files")
```
Like above, I think we want to log the exception first.

zmoon replied:
Attempted to address in 3227263

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+
+    @staticmethod
+    @logit(logger)
+    def open_qfed(files: List[str], out_var_dict: Dict[str, str] = None) -> xr.Dataset:
```
This method looks very similar to `process_hfed()`. Can we combine them somehow?

zmoon replied:
Removed in ca3bc1a7f7

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+        ratio_interp = ratio.sel(lat=clim.lat, lon=clim.lon, method="nearest")
+
+        # Loop through each time slice and scale the climatology
+        for index in range(0, len(clim.time)):
+            # Get the current time slice of the climatology
+            clim_slice = clim.data[index, :, :]
+
+            # Scale the current time slice
+            scaled_slice = clim_slice * ratio_interp[index, :, :]
+
+            # Update the climatology with the scaled time slice
+            clim.data[index, :, :] = scaled_slice.squeeze().data
```
Asked previously if this could be replaced by this instead of looping:
```suggestion
        # Scale data
        clim.data = clim.data * ratio_interp
```

zmoon replied:
I assume it could but I'll need to check with @bbakernoaa

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,427 @@ def __init__(self, config: Dict[str, Any]) -> None:
+                            dset[v].data = (
+                                ratio * dset[v] + (1 - ratio) * climo_scaled[v].data[tslice, :, :]
+                            )
+                        else:
+                            dset[v] = dset[v]
```
Setting this slice equal to itself doesn't do anything.
```suggestion
```

WalterKolczynski-NOAA reviewed (changes requested):
Let's set up a meeting early next week to discuss so we can get this finished. Just find a time on my calendar. @zmoon @bbakernoaa

zmoon committed:
Does nothing

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

zmoon committed:
Log before raise

and `logger.exception()` messages in `except` don't need to include the
error message, since the output automatically includes the exception
type, message, and traceback

zmoon committed:
Fix HFED guard in init

zmoon committed:
Updates following convo with Barry

zmoon committed:
Dict of files from yaml

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

zmoon committed:
Use new files dict

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -260,14 +211,14 @@ def open_climatology(files: List[str]) -> xr.Dataset:
         logger.info("Process Climatology Files")
-        for filename in sorted(files):
-            logger.info(f"  Opening Climatology File: {filename}")
+        for f in sorted(files):
```
Why did you remove the descriptive loop variable and replace it with a single character?

zmoon replied:
For consistency with the `open_qfed` method.

I don't like "filename" since it's really a path (name), not just a file name. I could go for, e.g., `p`/`fp`/`path`/`file` instead if you'd like.

WalterKolczynski-NOAA replied:
I'd be fine with `filepath`.

zmoon replied:
Addressed in 46fdc0bba3b8150466ebd67f748bd27b633124c7

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -233,9 +184,9 @@ def open_qfed(files: List[str], out_var_dict: Dict[str, str] = None) -> xr.Datas
-                logger.exception(f"FATAL ERROR: unable to read dataset {ee}")
-                raise Exception("FATAL ERROR: Unable to read dataset, ABORT!")
+            except Exception as e:
+                logger.exception(f"Unable to read dataset: {f}")
```
Even though it may seem a bit redundant with an uncaught exception, I think we want to keep "FATAL ERROR" in the logger (NCO will want to see it).
```suggestion
                logger.exception(f"FATAL ERROR: Unable to read dataset: {f}")
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -260,14 +211,14 @@ def open_climatology(files: List[str]) -> xr.Dataset:
-                logger.exception("Encountered an error reading climatology file, {error}".format(error=ee))
-                raise Exception("FATAL ERROR: Unable to read file, ABORT!")
+            except Exception as e:
+                logger.exception(f"Encountered an error reading climatology file: {f}")
```
```suggestion
                logger.exception(f"FATAL ERROR: Encountered an error reading climatology file: {f}")
```

WalterKolczynski-NOAA commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -301,28 +252,28 @@ def write_ncf(dset: xr.Dataset, outfile: str) -> None:
-            logger.exception("Encountered an exception in writing dataset, {}".format(ee))
-            raise Exception("FATAL ERROR: Unable to write dataset, ABORT!")
+        except Exception as e:
+            logger.exception("Encountered an error writing dataset")
```
```suggestion
            logger.exception("FATAL ERROR: Encountered an error writing dataset")
```

WalterKolczynski-NOAA reviewed (changes requested):
Just pretty small stuff, but there is also a merge conflict to resolve with develop.

zmoon committed:
FATAL ERROR

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

zmoon committed:
Merge remote-tracking branch 'origin/develop' into feature/fire_blending

zmoon committed:
`f` -> `filepath`

WalterKolczynski-NOAA reviewed (dismissed):
Still awaiting the new ExtData and any changes associated with that.

zmoon commented on ush/python/pygfs/task/aero_emissions.py:
```diff
@@ -37,48 +38,363 @@ def __init__(self, config: Dict[str, Any]) -> None:
+            }
         )
 
+        # TODO: if AERO_EMIS_FIRE is not 'blended' we don't want to do anything!
```
@WalterKolczynski-NOAA could you share some insights on how we should accomplish this? Is there a way to get the value of `AERO_EMIS_FIRE` (from `parm/config/{gefs,gfs)/config.aero`) within this Task? Or should the Task just be skipped somewhere further up the workflow scripting chain?

cc: @bbakernoaa

WalterKolczynski-NOAA replied:
If the job isn't needed at all, then it should be skipped at the workflow level (in `gfs_tasks.py`/`gefs_tasks.py`).

WalterKolczynski-NOAA replied:
Sorry, it should be removed in `workflow/applications/gfs_cycled.py` etc, which may also require changed to `gfs_tasks.py`, etc.

zmoon committed:
Merge remote-tracking branch 'origin/develop' into feature/fire_blending

WalterKolczynski-NOAA commented:
@bbakernoaa @zmoon This PR has been sitting for a long time. Any update on when a new ExtData will be available?

bbakernoaa commented:
> @bbakernoaa @zmoon This PR has been sitting for a long time. Any update on when a new ExtData will be available?

You can find the new ExtData directory here: `/scratch1/RDARCH/rda-arl-gpu/Barry.Baker/emissions/GEFS_ExtData`

bbakernoaa commented:
> > @bbakernoaa @zmoon This PR has been sitting for a long time. Any update on when a new ExtData will be available?
> 
> You can find the new ExtData directory here: `/scratch1/RDARCH/rda-arl-gpu/Barry.Baker/emissions/GEFS_ExtData`

@aerorahul @WalterKolczynski-NOAA

WalterKolczynski-NOAA commented:
> > > @bbakernoaa @zmoon This PR has been sitting for a long time. Any update on when a new ExtData will be available?
> > 
> > 
> > You can find the new ExtData directory here: `/scratch1/RDARCH/rda-arl-gpu/Barry.Baker/emissions/GEFS_ExtData`
> 
> @aerorahul @WalterKolczynski-NOAA

#3394

WalterKolczynski-NOAA commented:
@bbakernoaa @zmoon Please fill in the PR template and resolve the conflicts while we work on getting the new data in place.

bbakernoaa commented:
@zmoon check out the conflicts here please

zmoon committed:
Merge branch 'develop' into feature/fire_blending

bbakernoaa committed:
Merge branch 'develop' into feature/fire_blending

KateFriedman-NOAA commented:
@bbakernoaa @zmoon Checking in, what is the status of this? Perhaps can we close this and you can reopen when ready? Thanks!

bbakernoaa commented:
Closing due to the change with GCAFS priorities
